### PR TITLE
Ignore Git error messages in parsing

### PR DESCRIPTION
**Pull request by gerhardol**
Part of #7902 
https://github.com/gitextensions/gitextensions/issues/7902#issuecomment-604530322

## Proposed changes
This corrects the error parsing in git-diff and git-status (two implementations, git-diff and git-status before 2.11 shares parsing, git-status >= Git 2.11 separate).
Error messages are incorrectly parsed and interpreted as files.

The second commit presents the Git errors as special GitItemStatus that are presented 
(Without this, the problem is _less_ visible when the error messages are parsed correctly.)
It is difficult to transport the error from the parsing, to FileStatusList to the form (like RevisionDiff) .

Clarification:
 * git-diff for all Git versions uses GetDiffChangedFilesFromString/GetAllChangedFilesFromString_v1
 * git-status for Git = 2.11 uses GetAllChangedFilesFromString_v2
 * git-status for Git < 2.11 (unsupported but not blocked) uses GetAllChangedFilesFromString_v1

The PR is mostly opened for discussion

## Test methodology

Tests updated
----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
Are these corrections a for specific got versions? It just that mention got
2.11...

**Comment by gerhardol** [comment]
> Are these corrections a for specific got versions? It just that mention got
2.11...

All Git versions, but the code path is different for git-status < 2.11
The parsing is broken right now

**Comment by RussKie** [comment]
> All Git versions, but the code path is different for git-status < 2.11
> The parsing is broken right now

We min version we support 2.11. 
Once we release, I'd move a motion to set the min to 2.19.

File: GitCommands/Git/GitCommandHelpers.cs [hunk 1]
```diff
@@ -668,6 +668,30 @@ public static IReadOnlyList<GitItemStatus> GetStatusChangedFilesFromString(IGitM
             }
         }
 
+        private static string removeWarnings(string statusString)
```
Comment (RussKie) [inline GitCommands/Git/GitCommandHelpers.cs:671]:
```suggestion
        private static string RemoveWarnings(string statusString)
```

Comment (gerhardol) [reply GitCommands/Git/GitCommandHelpers.cs]: Done, squashed, rebased

**Comment by gerhardol** [comment]
> We min version we support 2.11.
Once we release, I'd move a motion to set the min to 2.19.

I agree for min version (but there are no major reason from GE perspective to force newer Git version, the git-status --porcelain=2 in 2.11 is the most breaking change we have)
We should probably update the recommended version, currently 2.23 to 2.26 (released 11 days ago)

Updated description for Git versions

I believe this PR is safe but the correction will hide the core errors why some other presentation is needed too, see description.

**Comment by gerhardol** [comment]
Rebased (some conflicts), this could be merged now but fit better after next release

File: GitCommands/Git/GitItemStatus.cs [hunk 2]
```diff
@@ -46,6 +46,7 @@ private enum Flags
 
         public string Name { get; set; }
         public string OldName { get; set; }
+        public string ErrorMsg { get; set; }
```
Comment (RussKie) [inline GitCommands/Git/GitItemStatus.cs:49]:
```suggestion
        public string ErrorMessage { get; set; }
```

File: GitCommands/Git/GitModule.cs [hunk 3]
```diff
@@ -2441,7 +2442,17 @@ public IReadOnlyList<GitItemStatus> GetDiffFiles(string firstRevision, string se
                 },
                 cache: noCache ? null : GitCommandCache);
 
-            var resultCollection = GitCommandHelpers.GetDiffChangedFilesFromString(this, output, stagedStatus).ToList();
+            var result = GitCommandHelpers.GetDiffChangedFilesFromString(this, output, stagedStatus).ToList();
+
+            if (IsGitErrorMessage(output))
+            {
+                // No simple way to pass the error message, create fake file
+                result.Add(new GitItemStatus()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2450]:
```suggestion
                result.Add(new GitItemStatus
```

File: GitCommands/Git/GitModule.cs [hunk 4]
```diff
@@ -2526,6 +2537,15 @@ public IReadOnlyList<GitItemStatus> GetTreeFiles(ObjectId commitId, bool full)
                 GitCommandHelpers.GetAllChangedFilesCmd(excludeIgnoredFiles, untrackedFiles));
 
             var result = GitCommandHelpers.GetStatusChangedFilesFromString(this, output).ToList();
+            if (IsGitErrorMessage(output))
+            {
+                // No simple way to pass the error message, create fake file
+                result.Add(new GitItemStatus()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2543]:
```suggestion
                result.Add(new GitItemStatus
```

File: GitCommands/Git/GitModule.cs [hunk 5]
```diff
@@ -2622,12 +2642,34 @@ public IReadOnlyList<GitItemStatus> GetIndexFiles()
...
                 output = _gitExecutable.GetOutput(command);
 
-                return GitCommandHelpers.GetStatusChangedFilesFromString(this, output)
+                var res = GitCommandHelpers.GetStatusChangedFilesFromString(this, output)
                     .Where(item => item.Staged == StagedStatus.Index)
                     .ToList();
+                if (!IsGitErrorMessage(output))
+                {
+                    // No simple way to pass the error message, create fake file
+                    res.Add(new GitItemStatus()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2651]:
```suggestion
                    res.Add(new GitItemStatus
```

File: GitCommands/Git/GitModule.cs [hunk 6]
```diff
@@ -2622,12 +2642,34 @@ public IReadOnlyList<GitItemStatus> GetIndexFiles()
...
+                }
+
+                return res;
+            }
+
+            var result = GitCommandHelpers.GetDiffChangedFilesFromString(this, output, StagedStatus.Index).ToList();
+            if (!IsGitErrorMessage(output))
+            {
+                // No simple way to pass the error message, create fake file
+                result.Add(new GitItemStatus()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2665]:
```suggestion
                result.Add(new GitItemStatus
```

File: GitCommands/Git/GitModule.cs [hunk 7]
```diff
@@ -2651,7 +2693,18 @@ public IReadOnlyList<GitItemStatus> GitStatus(UntrackedFilesMode untrackedFilesM
         {
             var args = GitCommandHelpers.GetAllChangedFilesCmd(true, untrackedFilesMode, ignoreSubmodulesMode);
             var output = _gitExecutable.GetOutput(args);
-            return GitCommandHelpers.GetStatusChangedFilesFromString(this, output);
+            var result = GitCommandHelpers.GetStatusChangedFilesFromString(this, output).ToList();
+            if (!IsGitErrorMessage(output))
+            {
+                // No simple way to pass the error message, create fake file
+                result.Add(new GitItemStatus()
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2700]:
```suggestion
                result.Add(new GitItemStatus
```

File: GitUI/GitUIExtensions.cs [hunk 8]
```diff
@@ -51,6 +51,12 @@ public static class GitUIExtensions
             [NotNull] string defaultText = "",
             [CanBeNull] Action openWithDiffTool = null)
         {
+            if (file != null && file.ErrorMsg.IsNotNullOrWhitespace())
```
Comment (RussKie) [inline GitUI/GitUIExtensions.cs:54]: I believe we're moving away from these extensions
```suggestion
            if (!string.IsNullOrWhitespace(file?.ErrorMsg))
```

File: GitCommands/Git/GitModule.cs [hunk 9]
```diff
@@ -2441,7 +2442,17 @@ public IReadOnlyList<GitItemStatus> GetDiffFiles(string firstRevision, string se
...
 
-            var resultCollection = GitCommandHelpers.GetDiffChangedFilesFromString(this, output, stagedStatus).ToList();
+            var result = GitCommandHelpers.GetDiffChangedFilesFromString(this, output, stagedStatus).ToList();
+
+            if (IsGitErrorMessage(output))
+            {
+                // No simple way to pass the error message, create fake file
+                result.Add(new GitItemStatus()
+                {
+                    Name = GitError,
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2452]: Why do we need to set the name?

File: GitCommands/Git/GitCommandHelpers.cs [hunk 10]
```diff
@@ -684,6 +684,30 @@ public static IReadOnlyList<GitItemStatus> GetStatusChangedFilesFromString(IGitM
             }
         }
 
+        private static string RemoveWarnings(string statusString)
```
Comment (RussKie) [inline GitCommands/Git/GitCommandHelpers.cs:687]: Is this tested via `TestGetStatusChangedFilesFromString`?

File: GitCommands/Git/GitItemStatus.cs (hunk 2 above)
Comment (gerhardol) [reply GitCommands/Git/GitItemStatus.cs]: ok

File: GitCommands/Git/GitModule.cs (hunk 3 above)
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs]: ok for all

File: GitUI/GitUIExtensions.cs (hunk 8 above)
Comment (gerhardol) [reply GitUI/GitUIExtensions.cs]: ok

File: GitCommands/Git/GitModule.cs (hunk 9 above)
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2452]: presented in FileStatusList

File: GitCommands/Git/GitCommandHelpers.cs (hunk 10 above)
Comment (gerhardol) [reply GitCommands/Git/GitCommandHelpers.cs:687]: yes, I updated the testcases, some existed before too

File: GitCommands/Git/GitModule.cs [hunk 11]
```diff
@@ -27,6 +27,7 @@ namespace GitCommands
     [DebuggerDisplay("GitModule ( {" + nameof(WorkingDir) + "} )")]
     public sealed class GitModule : IGitModule
     {
+        public static readonly string GitError = "Git Error";
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:30]: I don't think it needs to be public or static.
```suggestion
        private const string GitError = "Git Error";
```

Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:30]: OK (used in other modules initially, but creation of GitItemStatus should maybe be centralised anyway).

**Comment by gerhardol** [comment]
Should we merge this now?
This is not a prio for 4.0 (I submitted it initially more for discussion after #7902) but it feels safe (even if implementation details could be changed) and good to reduce open PRs.

**Comment by RussKie** [comment]
> Should we merge this now?

If it is ready - yes. Looks good.

**Comment by gerhardol** [comment]
> If it is ready - yes. Looks good.

Changed a few inverted checks, but I believe it is ready, with a limitation:
The error message is not always shown (Form Commit requires changes to ViewCurrentChanges() aswell as StagedStatus is set for the fake item (no!) or special handling in FormCommit (not nice either), most other handling of git-diff and git-status should be handled though since #7926)

But the current handling is incomplete.

**Comment by RussKie** [comment]
> But the current handling is incomplete.

Please either add TODOs where appropriate, or raise new issues to track what needs to be done.

**Comment by gerhardol** [comment]
After the discussion in #7978, I implemented the error handling in this PR too
Will squash and prepare to merge after yet another review

**Comment by gerhardol** [comment]
@msftbot merge in 24 h

**Comment by gerhardol** [comment]
@msftbot merge in 24 hours

**Pull request merged**